src/py21cmfast/src/Stochasticity.c

```diff
@@ -401,121 +350,90 @@ int fix_mass_sample(gsl_rng * rng, double exp_M, int *n_halo_pt, double *M_tot_p
             M_out[random_idx] = last_M_del;
             // LOG_ULTRA_DEBUG("kept.");
             *M_tot_pt += last_M_del;
-            n_removed--;
         }
     }
-    return n_removed;
-    //Possible future improvements
-    //  Try different target / trigger, which could overcome the M > exp_M issues better than a delta
-    //  e.g: set trigger for the check at exp_M - (M_Max - exp_M), target at exp_M
-    //  This provides a range of trigger points which can be absorbed into the below-resolution values
-    //  Obviously breaks for exp_M < 0.5 M, think of other thresholds.
-
 }
 
 /* Creates a realisation of halo properties by sampling the halo mass function and
  * conditional property PDFs, Sampling is done until there is no more mass in the condition
- * Stochasticity is ignored below a certain mass threshold*/
+ * Stochasticity is ignored below a certain mass threshold */
 int stoc_mass_sample(struct HaloSamplingConstants * hs_constants, gsl_rng * rng, int *n_halo_out, float *M_out){
-    //lnMmin only used for sampling, apply factor here
-    double mass_tol = global_params.STOC_MASS_TOL;
     double exp_M = hs_constants->expected_M;
-    if(hs_constants->from_catalog)exp_M *= 1.; //fudge factor for assuming that internal lagrangian volumes are independent
 
-    int n_halo_sampled, n_failures=0;
+    //fudge factor for assuming that internal lagrangian volumes are independent
+    exp_M *= user_params_stoc->HALOMASS_CORRECTION;
```

Comment:
this comment and code are unclear to me

Reply:
This is the implementation of the above correction in the mass-limited sampling. One of the approximations in our sample is that we sample the final CMF in an uncorrelated way, assuming each sampled halo is independent of the others. This means that the CMF doesn't change after each sample, like it does in the partition or binary split methods. This results in a bias where the last halo sampled is on average larger.

I'll change the comment to something more generic and expand the explanation in `inputs.py`

Reply:
Thanks!